app.js

```diff
@@ -1,54 +1,231 @@
+    {
+      id: "ad_-OwhQA",
+      name: "apple",
+      price: "2.99"
+    },
+    {
+      id: "DLU6Rjfon",
+      name: "banana",
+      price: "1.99"
+    },
+    {
+      id: "3YVN-Kn_z",
+      name: "kiwis",
+      price: "4.13"
+    },
+    {
+      id: "0BtTp4JRn",
+      name: "melon",
+      price: "3.33"
+    }
+  ],
+  vegetables: [
+    {
+      id: "xg7hs7vlV",
+      name: "acorn squash",
+      price: "1.33"
+    },
+    {
+      id: "0k8RBDjm0",
+      name: "artichoke",
+      price: "1.23"
+    },
+    {
+      id: "Vt2Y8l3IE",
+      name: "basil",
+      price: "1.93"
+    },
+    {
+      id: "-iWrxvN8G",
+      name: "cabbage",
+      price: "2.98"
+    }
+  ]
+};
 
-app.get('/vegetables', (req, res, next) => {
-  const { vegetables } = data
-  res.json(vegetables)
-})
+app.get("/vegetables", (req, res, next) => {
+  const { vegetables } = data;
 
-app.get('/vegetables/:id', (req, res, next) => {
-  const { vegetables } = data
-  const { id } = req.params
-  const vegetable = vegetables.find(veggie => veggie.id === id)
+  const name = req.query.name;
+
+  let myList = [];
+
+  if (!name) {
+    return res.json(vegetables);
+  }
+  console.log(name);
+
+  myList = vegetables.filter(veggie =>
+    veggie.name.includes(name.substring(1, req.query.name.length - 1))
+  );
+
+  res.json(myList);
+});
+/******************************************** */
+app.get("/vegetables/:id", (req, res, next) => {
+  const { vegetables } = data;
+  const id = req.params.id;
+  let vegetable = vegetables.find(veggie => veggie.id === id);
+
+  console.log(id);
+
+  if (!vegetable) {
+    const message = `Could not find vegetable with ID of ${id}`;
+    next({ status: 404, message });
+  }
+
+  res.json(vegetable);
+});
+/******************************************** */
+app.delete("/vegetables/:id", (req, res, next) => {
+  const { vegetables } = data;
+  const { id } = req.params;
+  let vegetable = vegetables.find(veggie => veggie.id === id);
+
+  //console.log(id);
+
+  if (!vegetable) {
+    const message = `DELETE: Could not find vegetable with ID of ${id}`;
+    next({ status: 404, message });
+  }
+  let filtered = vegetables.filter(
+    (x = (value, index, arr) => {
+      return value.id !== id;
+    })
+  );
+  //console.log(data.vegetables);
+  //console.log("===========================");
+  data.vegetables = filtered;
+  //console.log(data.vegetables);
+  res.status(200).json(vegetable);
+});
+/******************************************** */
+app.put("/vegetables/:id", helpers.validate, (req, res, next) => {
+  const { vegetables } = data;
+  const { id } = req.params;
+  const { name, price } = req.body;
+  const vegetable = vegetables.find(veggie => veggie.id === id);
 
   if (!vegetable) {
-    const message = `Could not find vegetable with ID of ${id}`
-    next({ status: 404, message })
+    const message = `PUT: Could not find ${name} with ID of ${id}`;
+    next({ status: 404, message });
   }
+  vegetable.name = name;
```

Comment:
I'm getting an error here of `TypeError: Cannot set property 'name' of undefined`. My guess is you need to `return` here to stop the rest of the function from executing.